### DISCUSSION TITLE
Optimisations to the Skript Parser

### DISCUSSION THREAD

**Discussion by bluelhf** [SkriptLang/Skript#4598] [Suggestions]
A common issue with Skript's parsing system is how long, complicated lines cause the server to freeze (and ultimately stop due to the server watchdog). Perhaps it'd be a good idea to start work on optimising parsing?

- IDEA has built-in capabilities for performance profiling; we could likely isolate parsing of complicated code and find common bottlenecks. The complexity of possible optimisations would depend on the nature of the bottlenecks we encounter. 
- Server crashes could be avoided by using a dedicated, asynchronous thread for parsing Skript code. The parser does have some functionality that may become problematic, for instance the script load events that may be triggered during parsing (those would still need to be ran on the main thread, perhaps by blocking the parser thread or by deferring execution until parsing is complete).

**Comment by TPGamesNL** [comment]
Most of these things have at least partially been done or tried:
- For parser optimizations, PRs like #4176, #4493, #4275 and #4260 have been created for various parsing optimizations.
- Async and parallel script loading have been a thing for some time now, although they definitely have issues (IIRC functions don't work well with async loading at the moment), if this becomes more stable I think it'd be a good idea to make async loading the default.

**Reply by bluelhf** [reply]
Ah yes, I'm aware of your excellent work optimising the parser :)

I noticed the problems with asynchronous and parallel script loading in my testing, too. I believe it's not just limited to functions, but there are issues also with unsynchronised collections and such. Thread safety in general seems to be an area Skript handles particularly badly. Perhaps this is something to focus more on in the future, both through code reviews and refactoring?

**Reply by TPGamesNL** [reply]
There are probably plenty more issues with async loading, but the problem is that we currently have little reports about them (if you found any issues please report \:) )
